### PR TITLE
CLI 0.4.1

### LINKED ISSUE #2663: Use the REST API in the CLI when applicable
The metastore should send events to the control plane so that a new source is immediately detected.

This can be done if the CLI uses a gRPC metastore or the REST API.

Currently, the CLI is using a quickwit config file but we just need to have an instance address if we use the REST API or the metastore address if we use the gRPC metastore.

### PR DESCRIPTION
**Pull request by fmassot**
Fix #2663 

## TODO for this PR:

- [x] Add ingest API CLI
- [x] Use splits endpoint with filters endpoint
- [x] Use `mark_splits_for_deletion` endpoint
- [x] Use `reset_source_checkpoint`  endpoint
- [x] Use toggle source  endpoint
- [x] Use clear index endpoint
- [x] use dry run delete index endpoint
- [x] use create index with overwrite  endpoint

## TODO later:

- Add `quickwit-integration` to isolate these tests (next PR before release 0.4.1)
- Add tests on CLI commands `mark_splits_for_deletion`, `toggle-source`, `reset source checkpoint`, `clear-index` (next PR before release 0.4.1)
- Update CLI docs (next PR before release 0.4.1)
- Add tags to splits filter endpoint: remove the tags args for the 0.4.1 and open an issue to add that later, not a high priority
- Display files sizes in describe split command, not a high priority
- the REST client depends on quickwit-server because we use `ListSplitsQueryParams` and `SearchRequestQueryString`. We should either move these struct in `quickwit-common` or having separated struct for the client.
- On a delete index, check that all nodes are doing a cleanup of the index directory.

## CLI preview

```bash
USAGE:
    quickwit [OPTIONS] <SUBCOMMAND>

OPTIONS:
    -h, --help        Print help information
        --no-color    Disable ANSI terminal codes (colors, etc...) being injected into the logging
                      output [env: NO_COLOR=]
    -V, --version     Print version information
    -y, --yes         Assume "yes" as an answer to all prompts and run non-interactively.

SUBCOMMANDS:
    run       Starts Quickwit server with all services by default (`indexer`, `searcher`...).
    index     Manages indexes: creates, deletes, ingests, searches, describes...
    source    Manages sources: creates, updates, deletes sources...
    split     Manages splits: lists, describes, marks for deletion...
    tool      Performs utility operations. Requires a node config.
```

### CODE REVIEW

File: quickwit/quickwit-cli/src/split.rs [hunk 1]
```diff
@@ -332,32 +335,24 @@ impl SplitCliCommand {
...
+    let transport = Transport::new(args.cluster_endpoint);
+    let qw_client = QuickwitClient::new(transport);
+    let list_splits_query_params = ListSplitsQueryParams {
+        split_states: args.split_states,
+        start_timestamp: args.start_date.map(OffsetDateTime::unix_timestamp),
+        end_timestamp: args.end_date.map(OffsetDateTime::unix_timestamp),
+        end_create_timestamp: args.create_date.map(OffsetDateTime::unix_timestamp),
+    };
+    // TODO: plug tags.
+    // if let Some(tags) = args.tags {
```
Comment (evanxg852000) [inline quickwit/quickwit-cli/src/split.rs:348]: maybe we can create the ticket and reference it here

File: quickwit/quickwit-cli/src/split.rs [hunk 2]
```diff
@@ -418,42 +405,44 @@ async fn describe_split_cli(args: DescribeSplitArgs) -> anyhow::Result<()> {
...
-                file_name: path.to_str().unwrap().to_string(),
-                size: format_size(size, DECIMAL),
-            });
-        }
-        let hotcache_table = make_table("Files in Hotcache", hotcache_files.into_iter(), false);
-        println!("{hotcache_table}");
-    }
+    println!("{}", make_split_table(&[split], "Split"));
+
+    // TODO: if we have access to the storage, we could fetch that.
```
Comment (evanxg852000) [inline quickwit/quickwit-cli/src/split.rs:410]: Maybe a sign that we need to persist the size of the split in the metastore when registering the split.

File: quickwit/quickwit-cli/tests/cli.rs [hunk 3]
```diff
@@ -83,49 +83,67 @@ fn test_cmd_help() {
...
     let index_id = append_random_suffix("test-create-cmd");
     let test_env = create_test_env(index_id, TestStorageType::LocalFileSystem).unwrap();
+    test_env.start_server().await.unwrap();
     create_logs_index(&test_env).await.unwrap();
 
     let index_metadata = test_env.index_metadata().await.unwrap();
     assert_eq!(index_metadata.index_id(), test_env.index_id);
 
-    // Create without giving `index-uri`.
+    // Create on existing index should fail.
```
Comment (evanxg852000) [inline quickwit/quickwit-cli/tests/cli.rs:95]:
```suggestion
    //Creating an existing index should fail.
```

File: quickwit/quickwit-cli/tests/cli.rs [hunk 4]
```diff
@@ -83,49 +83,67 @@ fn test_cmd_help() {
...
 
     let index_metadata = test_env.index_metadata().await.unwrap();
     assert_eq!(index_metadata.index_id(), test_env.index_id);
     assert_eq!(index_metadata.index_uri(), &test_env.index_uri);
+}
 
-    // Create non existing index with --overwrite.
-    let index_id = append_random_suffix("test-create-non-existing-index-with-overwrite");
-    let test_env = create_test_env(index_id, TestStorageType::LocalFileSystem).unwrap();
+// #[tokio::test]
```
Comment (evanxg852000) [inline quickwit/quickwit-cli/tests/cli.rs:124]: This can be uncommented now & updated

File: quickwit/quickwit-cli/tests/cli.rs [hunk 5]
```diff
@@ -258,12 +235,33 @@ async fn test_ingest_docs_cli() {
     let cache_directory_path = get_cache_directory_path(&test_env.data_dir_path);
 
     assert!(cache_directory_path.read_dir().unwrap().next().is_none());
+
+    // Ingest on non existing file should fail.
```
Comment (evanxg852000) [inline quickwit/quickwit-cli/tests/cli.rs:239]:
```suggestion
    // Ingest a non-existing file should fail.
```

File: quickwit/quickwit-cli/tests/cli.rs [hunk 6]
```diff
@@ -449,100 +443,69 @@ async fn test_search_index_cli() {
...
-
-    assert!(!test_env
-        .data_dir_path
-        .join(INDEXING_DIR_NAME)
-        .join(test_env.index_id)
-        .as_path()
-        .try_exists()
-        .unwrap());
-}
+// #[tokio::test]
```
Comment (evanxg852000) [inline quickwit/quickwit-cli/tests/cli.rs:446]: Maybe  a skip  attribute `#[ignore]`

**Review by evanxg852000 — APPROVED** [review]

File: quickwit/quickwit-cli/tests/cli.rs (hunk 4 above)
Comment (fmassot) [reply quickwit/quickwit-cli/tests/cli.rs:124]: yes! well spotted

File: quickwit/quickwit-cli/tests/cli.rs (hunk 6 above)
Comment (fmassot) [reply quickwit/quickwit-cli/tests/cli.rs:446]: Let's  keep this test, I added the `dry_run` parameter in the REST API.

File: quickwit/quickwit-cli/src/split.rs (hunk 1 above)
Comment (fmassot) [reply quickwit/quickwit-cli/src/split.rs:354]: will do

File: quickwit/quickwit-cli/src/split.rs (hunk 2 above)
Comment (fmassot) [reply quickwit/quickwit-cli/src/split.rs:425]: We have it already in the metadata; I'm going to open an issue to track this.

**Pull request merged**